### ISSUE TITLE
[Windows/ARM64] Unable to build content

### ISSUE THREAD

**Issue by airtaxi** [MonoGame/MonoGame#8398]
### Prerequisites

- [X] I have verified this issue is present in the `develop` branch
- [X] I have searched [open](https://github.com/MonoGame/MonoGame/issues) and [closed](https://github.com/MonoGame/MonoGame/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported.

### MonoGame Version

MonoGame 3.8.1.303

### Which MonoGame platform are you using?

MonoGame Cross-Platform Desktop Application (mgdesktopgl)

### Operating System

Windows 11 24H2 (26100.994)

### Description

Two major issues have been identified on the ARM version of Windows:
1. `FreeImage.dll` cannot be loaded.
2. Compilation of `.fx` files results in format-related errors.

### Issue 1: FreeImage.dll Loading Failure

- **Affected File**: `FreeImageAPI.cs`
- **Error Location**: Line 120, related to loading `FreeImage.dll`
- **Error Message**:
  ```
  D:/Repos/LemegetonDiary/LemegetonDiary/Content/Menu/Guide/KeyIcon/MouseL.png : error : Importer 'TextureImporter' had unexpected failure! [D:\Repos\LemegetonDiary\LemegetonDiary\LemegetonDiary.csproj]
    System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
       at FreeImageAPI.FreeImage.GetFileTypeU(String filename, Int32 size)
       at FreeImageAPI.FreeImage.GetFileType(String filename, Int32 size) in /home/runner/work/MonoGame/MonoGame/MonoGame.Framework.Content.Pipeline/Utilities/FreeImageAPI.cs:line 120
       at Microsoft.Xna.Framework.Content.Pipeline.TextureImporter.Import(String filename, ContentImporterContext context) in /home/runner/work/MonoGame/MonoGame/MonoGame.Framework.Content.Pipeline/TextureImporter.cs:line 83
       at Microsoft.Xna.Framework.Content.Pipeline.ContentImporter`1.Microsoft.Xna.Framework.Content.Pipeline.IContentImporter.Import(String filename, ContentImporterContext context) in /home/runner/work/MonoGame/MonoGame/MonoGame.Framework.Content.Pipeline/ContentImporter.cs:line 45
       at MonoGame.Framework.Content.Pipeline.Builder.PipelineManager.ProcessContent(PipelineBuildEvent pipelineEvent) in /home/runner/work/MonoGame/MonoGame/MonoGame.Framework.Content.Pipeline/Builder/PipelineManager.cs:line 669
  ```

#### FreeImage.dll Information
- **Path**: `C:\Users\kck41\.nuget\packages\dotnet-mgcb\3.8.1.303\tools\net6.0\any\FreeImage.dll`
- **Dump Details**:
  ```
  Microsoft (R) COFF/PE Dumper Version 14.40.33811.0
  Copyright (C) Microsoft Corporation.  All rights reserved.

  Dump of file C:\Users\kck41\.nuget\packages\dotnet-mgcb\3.8.1.303\tools\net6.0\any\FreeImage.dll

  PE signature found

  File Type: DLL

  FILE HEADER VALUES
              8664 machine (x64)
                 7 number of sections
          5F02481F time date stamp Mon Jul  6 06:37:35 2020
                 0 file pointer to symbol table
                 0 number of symbols
                F0 size of optional header
              2022 characteristics
                     Executable
                     Application can handle large (>2GB) addresses
                     DLL
  ...
  ...
  ```

### Issue 2: .fx File Compilation Error

- **Error Message**:
  ```
  D:/Repos/LemegetonDiary/LemegetonDiary/Content/Effects/BlendMode.fx: An attempt was made to load a program with an incorrect format. (0x8007000B)
  Unexpected error compiling 'D:/Repos/LemegetonDiary/LemegetonDiary/Content/Effects/BlendMode.fx'!
  ```

### Device Information
- **Model**: [ASUS Vivobook S 15 OLED](https://www.asus.com/laptops/for-home/vivobook/asus-vivobook-s-15-s5507/)
- **CPU**: Snapdragon(R) X Elite - X1E78100
- **Memory**: 32GB

### Additional Information
- The same build was tested on a x64-based device (Ryzen 9 7940HS) running Windows, and no issues were encountered.

### Steps to Reproduce

1. Attempt to build the project on an ARM version of Windows.
2. Observe the failure when `FreeImage.dll` is loaded.
3. Note the `System.BadImageFormatException` error.
4. Attempt to compile `.fx` files and observe similar format-related errors.

### Expected Behavior

- `FreeImage.dll` should load without errors, allowing the `FreeImage_GetFileTypeU` function to execute correctly.
- `.fx` files should compile successfully without format errors.

### Resulting Behavior

- A `System.BadImageFormatException` occurs, indicating an attempt to load a program with an incorrect format for `FreeImage.dll`.
- Compilation of `.fx` files fails with a similar format error.

### Files

- [Full Build Log](https://github.com/user-attachments/files/15949532/log.txt)

**Comment by airtaxi** [comment]
### Additional Information

After building FreeImage.dll and libmojoshader_64.dll for ARM64, execution was possible. However, the compilation results of the shader (.fx) files did not behave as expected.

Build result on ARM64:
![image](https://github.com/MonoGame/MonoGame/assets/6309929/b1b7bce8-b2bd-4930-ab0f-5ec1fb4598fd)

Expected result:
![image](https://github.com/MonoGame/MonoGame/assets/6309929/4086dda8-abaa-434a-adff-17a980000252)


For mojoshader, the exact source code used was unknown, so the build from the following URL was used for ARM64:
[https://github.com/icculus/mojoshader/tree/aca99017596f92d1b9abe948371100f1c9409706](https://github.com/icculus/mojoshader/tree/aca99017596f92d1b9abe948371100f1c9409706)

If the source code of mojoshader used by MonoGame could be provided, it is anticipated that proper execution would be possible.

**Comment by mrhelmut** [comment]
```win-arm64``` is an RID that we currently haven't planned to support, for both the developer experience and runtime.

Unless someone is willing to contribute automated building of all dependencies to that target (like it has been done for other platforms, see the [MonoGame.Library.* projects](https://github.com/MonoGame)), the Foundation hasn't planned to incentivize it at the moment.

**Comment by airtaxi** [comment]
I think it would be beneficial to support building for x64 targets in an ARM64 environment, at the very least.
After replacing the `FreeImage.dll` file mentioned in the second comment and the `mojoshader` from the MonoGame repository with the version I built for ARM64 (renamed to `libmojoshader_64.dll`), I successfully built a game targeting x64.

![image](https://github.com/MonoGame/MonoGame/assets/6309929/3dd1d948-f176-4789-8898-e2d71f7bf4d5)

As initially reported, building for x64 targets in an ARM64 environment is currently impossible in both Visual Studio and Rider without replacing these two DLL files. Therefore, for the next version, I suggest the following improvements:

1. Replace the attached DLL files in the `dotnet-mgcb\3.8.1.303\tools\net6.0\any` directory for ARM64 devices.
2. Alternatively, add a specific suffix (such as `_arm64.dll`) to these DLL files so that the ARM64 environment targets these modified DLL files for supporting x64 builds on ARM64.

These changes would significantly enhance the development experience for ARM64 users by enabling seamless x64 builds.

[DLLs.zip](https://github.com/user-attachments/files/15965175/any.zip)

**Comment by mrhelmut** [comment]
It's not just about these dlls and there is more to it than what your project needs to have a proper experience on win-arm64.

Since win-arm64 isn't supported by Steam, it's a really low priority to support this scenario.

**Comment by airtaxi** [comment]
Sure. I understand that MonoGame currently has a lot of priorities on its roadmap. In that case, if possible, I will try to implement and contribute build automation for the ARM64 environment in my spare time if possible.

**Comment by mrhelmut** [comment]
That would be welcome.

**Comment by MCDFsteve** [comment]
Wow, this game looks awesome! Will it be available on Steam?

**Comment by airtaxi** [comment]
@MCDFsteve Yes, it will be. Thanks for the compliment!
The game name is "Lemegeton Diary" and is in development progress.
But it will be only available in Korea at least for now.

**Comment by MCDFsteve** [comment]
> @airtaxi Brilliant, thank you so much for saving my day!! :D
> 
> @mrhelmut You better revisit the stance to win-arm64 optimized games given the recent new wave of laptops with the Snapdragon Elite X. They are getting rave reviews, excellent battery life and performance, and is the competitive answer to Apples modern M-line chips. Market adoption will likely be very high.

It is not better than Apple's Mac in any aspect except for being more expensive. It is a product that combines the drawbacks of both Windows devices and Mac devices.

**Comment by airtaxi** [comment]
> except

I'm daily driving new Surface Laptop 7th edition with Snapdragon X Elite and I love it.
Coding and debugging game with Visual Studio only drains about 10% battery per hour and gives basically same performance compared when plugged in.

Also, it's more cheaper than baseline Macbook Air M3. And moreover, you can get cheaper device like Asus Vivobook S 15 OLED if you compromising build quality.

**Comment by airtaxi** [comment]
@sinsro I'm glad if I helped you :)
Also, not mentioned here, if you're using built-in MonoGame's font feature, you also need to read [this comment](https://github.com/MonoGame/MonoGame.Dependencies/pull/166#issuecomment-2264842434) since there is more .dll file to replace if you use that.

**Comment by MCDFsteve** [comment]
> > except
> 
> I'm daily driving new Surface Laptop 7th edition with Snapdragon X Elite and I love it. Coding and debugging game with Visual Studio only drains about 10% battery per hour and gives basically same performance compared when plugged in.
> 
> Also, it's more cheaper than baseline Macbook Air M3. And moreover, you can get cheaper device like Asus Vivobook S 15 OLED if you compromising build quality.

I plan to buy and use it; I really like the Surface series.

**Comment by mrhelmut** [comment]
I understand that some of you are excited about hardware and stuff, but here's the truth based on Steam's hardware survey (which is very relevant in regard to hardware market trends): ```win-arm64``` is 0.08% of Windows users and tends to decline over time.

There might be some enthusiasts and a tech bubble around it, but given those figures, the Foundation is likely to leave ARM64 support to the community.

Any contribution in that regard will be more than welcome, just don't expect incentives.

**Comment by MrValentine7777** [comment]
Just to point out, Windows on ARM is on its third or so wave now, until it reaches 8th generation I would ignore it if I were you. [It tends to die before the third generation comes out.]

Coming from a previous Gen Enthusiast.

And I do want one of the new WinARM devices but don't see the point as of yet.

**Comment by stromkos** [comment]
@MrValentine7777 

> It tends to die before the third generation comes out.

Die?

My Windows Phone 7 is sitting here on my desk. It runs XNA just fine :P

**Comment by MrValentine7777** [comment]
> desk

Trolling?

**Comment by MrValentine7777** [comment]
Still got my WP8/10 devices too though, best OS for visually impaired users, hands down, such a loss.

**Comment by stromkos** [comment]
I come from Windows CE times. The push for Windows 8 interface on desktop killed much of the momentum. :(

**Comment by MrValentine7777** [comment]
Unfortunately people don't like good things...

**Comment by KimPlaybit** [comment]
> @sinsro I'm glad if I helped you :)
> Also, not mentioned here, if you're using built-in MonoGame's font feature, you also need to read [this comment](https://github.com/MonoGame/MonoGame.Dependencies/pull/166#issuecomment-2264842434) since there is more .dll file to replace if you use that.

This seems to be having trouble, tells me "Method not found: 'Int32 SharpFont.SizeMetrics.get_Height()'.". where and how do i solve this?

**Comment by dellis1972** [comment]
> I understand that some of you are excited about hardware and stuff, but here's the truth based on Steam's hardware survey (which is very relevant in regard to hardware market trends): `win-arm64` is 0.08% of Windows users and tends to decline over time.
> 
> There might be some enthusiasts and a tech bubble around it, but given those figures, the Foundation is likely to leave ARM64 support to the community.
> 
> Any contribution in that regard will be more than welcome, just don't expect incentives.

I would like to point out that this issue is for developing games on windows arm, not playing games on windows arm. The Steam survey data will be biased towards people who play games. Developers tend to move to these new hardware platforms earlier , they are early adopters. 

@airtaxi have you been able to take a look at the links in https://github.com/MonoGame/MonoGame.Dependencies/pull/166#issuecomment-2199445040. I've  done a few win-arm64 PR's for SDL and OpenAL . But it would be helpful to have the content pipeline things like Assimp etc upgraded.

**Comment by ThomasFOG** [comment]
>  I've done a few win-arm64 PR's for SDL and OpenAL

Speaking of which, for the time being these changes are on their own (nameless) branch and not merged into the distributed nugets. We might want to move that forward so that ```DesktopGL``` games built with ```dotnet publish -r win-arm64``` pick up the correct binaries from the nuget. We would at least have games working without replacing dlls manually.

**Comment by dellis1972** [comment]
> > I've done a few win-arm64 PR's for SDL and OpenAL
> 
> Speaking of which, for the time being these changes are on their own (nameless) branch and not merged into the distributed nugets. We might want to move that forward so that `DesktopGL` games built with `dotnet publish -r win-arm64` pick up the correct binaries from the nuget. We would at least have games working without replacing dlls manually.

Why were they not distributed? the changes were merged last year.

**Comment by ThomasFOG** [comment]
Correction: they got merged into ```main``` (although the git history is misleading on this one because a tag has been put on a dead branch).

Although, it's incomplete, ```win-arm64``` is correctly being built, but the dll isn't packaged into the nuget.

**Comment by airtaxi** [comment]
**Final, up-to-date workaround that covers both assimp and font issues:**

1. Install the **x64 version** of the .NET SDK **8 or later**.
   This will not replace your existing ARM64 build installation, so it is safe to install alongside it.

2. Add the following to a `PropertyGroup` in your `.csproj` file:

   ```
   <DotnetCommand Condition="'$(OS)' == 'Windows_NT' And '$(PROCESSOR_ARCHITECTURE)' == 'ARM64'">
     "$(ProgramFiles)\dotnet\x64\dotnet.exe"
   </DotnetCommand>
   ```

3. If you previously applied the patch I posted earlier, delete the
   `%USERPROFILE%\.nuget\packages\dotnet-mgcb` folder.
   Then try building the project once so that the NuGet packages are restored again.
   The build is expected to fail at this stage.

4. Download the DLL file below and overwrite it in the following folder:
   `%USERPROFILE%\.nuget\packages\dotnet-mgcb\<version in use>\tools\net8.0\any`
[MonoGame.Framework.Content.Pipeline.zip](https://github.com/user-attachments/files/24928940/MonoGame.Framework.Content.Pipeline.zip)

After this, **x64 builds should succeed successfully**.

---

**If you want to build the game as a native ARM64 binary, you need to perform the additional steps below:**

1. Download the ZIP file below, create a folder named `win-arm64` in the same directory as your `.csproj` file, and place the two DLL files from the ZIP into that folder.
[win-arm64.zip](https://github.com/user-attachments/files/24928953/win-arm64.zip)

2. Add the following `ItemGroup` and targets to your `.csproj` file:

   ```
   <ItemGroup>
       <_ARM64Binaries Include="win-arm64\*.*" />
   </ItemGroup>

   <Target Name="CopyAfterBuild" AfterTargets="Build" Condition=" '$(Platform)' == 'ARM64' ">
       <Copy SourceFiles="@(_ARM64Binaries)" DestinationFolder="$(OutDir)" />
   </Target>

   <Target Name="CopyAfterPublish" AfterTargets="Publish" Condition=" '$(Platform)' == 'ARM64' ">
       <Copy SourceFiles="@(_ARM64Binaries)" DestinationFolder="$(PublishDir)" />
   </Target>
   ```

**Comment by airtaxi** [comment]
**If `MonoGame.Framework.Content.Pipeline.dll` is incompatible due to using a higher version or a fork of MonoGame, please follow the steps below:**

1. Clone the MonoGame repository **including all submodules**.

2. In the `ExternalTool.cs` file of the `MonoGame.Framework.Content.Pipeline` project, insert the following code **immediately after the first line of the `Run` method**:

   ```csharp
   if (command == "dotnet" && RuntimeInformation.OSArchitecture == Architecture.Arm64)
   {
       Console.WriteLine("Detected ARM64 architecture, attempting to use x64 dotnet.");
       var x64Dotnet = Path.Combine(
           Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles),
           "dotnet",
           "x64",
           "dotnet.exe");
       fullPath = FindCommand(x64Dotnet);
   }
   ```

   The result should look like the screenshot below.

<img width="1581" height="718" alt="Image" src="https://github.com/user-attachments/assets/669b92fe-4d79-47a5-b42f-9e86e36a2313" />

3. Build the `MonoGame.Framework.Content.Pipeline` project in **Release** mode, and use the resulting DLL file in the workaround described above.

**Comment by dellis1972** [comment]
@airtaxi  thanks for that info. 

I assume you have access to an ARM64 windows machine? If so if you have time it would be great to get some of these changes into the actual libraries , I made a start but testing is difficult as I don't have the required device.

https://github.com/MonoGame/MonoGame.Library.Assimp/pull/6

For reference these are all the other repositories which the content pipeline use that need an arm64 upgrade.

https://github.com/MonoGame/MonoGame.Library.MojoShader
https://github.com/MonoGame/MonoGame.Library.FreeImage
https://github.com/MonoGame/MonoGame.Library.FreeType
https://github.com/MonoGame/MonoGame.Tool.Crunch
https://github.com/MonoGame/MonoGame.Tool.BasisUniversal
https://github.com/MonoGame/MonoGame.Tool.ffprobe
https://github.com/MonoGame/MonoGame.Tool.FFmpeg
https://github.com/MonoGame/MonoGame.Tool.Dxc

All these repos use the same build scripts via cake. 

the following repositories already have win-arm64 builds, for some reason they are not appearing the in the nuget.

https://github.com/MonoGame/MonoGame.Library.SDL
https://github.com/MonoGame/MonoGame.Library.OpenAL

**Comment by dellis1972** [comment]
An update on this. 

The following have been updated. 

https://github.com/MonoGame/MonoGame.Library.MojoShader
https://github.com/MonoGame/MonoGame.Library.FreeImage
https://github.com/MonoGame/MonoGame.Library.FreeType
https://github.com/MonoGame/MonoGame.Library.Assimp
https://github.com/MonoGame/MonoGame.Library.SDL
https://github.com/MonoGame/MonoGame.Library.OpenAL
https://github.com/MonoGame/MonoGame.Tool.Dxc
https://github.com/MonoGame/MonoGame.Tool.Crunch
https://github.com/MonoGame/MonoGame.Tool.BasisUniversal

These need to be updated  windows-arm64 

https://github.com/MonoGame/MonoGame.Tool.ffprobe
https://github.com/MonoGame/MonoGame.Tool.FFmpeg

**Comment by airtaxi** [comment]
@dellis1972 
Hello, and thank you for the update on the current progress.

After building the develop branch of MonoGame, it appears that the ARM64 artifacts you mentioned are not yet available for testing. (In the build output of MonoGame.Content.Builder, the win-arm64 runtime does not seem to be present.)

Regarding contributing to the ARM64 build pipeline, I am currently unable to make direct contributions due to limited time while fulfilling mandatory military service in South Korea. However, I would be more than happy to assist with testing. Once testing is ready, please let me know. I can clone the develop branch, build it, and prepare a project that tests all available artifacts to analyze whether there are any issues with building and running.

I have two Snapdragon X Elite–based ARM64 devices available for testing (Surface Pro 12 inch and ASUS Vivobook S16 OLED).

**Comment by dellis1972** [comment]
So the only outstanding tools that need upgrading are ffmpeg, then we can start upgrading the main repo with the new versions.

**Comment by dellis1972** [comment]
If you are using the preview "Content Builder" project for content (rather than mgcb) you can replace the existing package references with these

```
    <PackageReference Include="MonoGame.Library.FreeType" Version="2.13.2.3" />
    <PackageReference Include="MonoGame.Library.MojoShader" Version="1.0.0.4" />
    <PackageReference Include="MonoGame.Tool.Basisu" Version="2.0.2" />
    <PackageReference Include="MonoGame.Tool.Crunch" Version="1.0.4.6" />
    <PackageReference Include="MonoGame.Tool.Dxc" Version="1.8.2505.11" />
```

This will upgrade. Note ffmpeg/ffprobe have not yet been updated.

**Comment by mirh** [comment]
https://github.com/MonoGame/MonoGame.Tool.FFprobe/pull/7 and https://github.com/MonoGame/MonoGame.Tool.FFmpeg/pull/9 landed too (and released). Are we good?

**Comment by dellis1972** [comment]
Closing this issue, this was fixed in 3.8.5 and above. If there are any other issues please open a new issue.
